uc-crux-llvm/src/UCCrux/LLVM/Constraints.hs

```diff
@@ -0,0 +1,451 @@
+-- compatibility.
+data Constraints m (argTypes :: Ctx (FullType m)) = Constraints
+  { _argConstraints :: Ctx.Assignment (ConstrainedShape m) argTypes,
+    _globalConstraints :: Map L.Symbol (Some (ConstrainedShape m)),
+    _relationalConstraints :: [RelationalConstraint m argTypes]
+  }
+
+argConstraints :: Simple Lens (Constraints m argTypes) (Ctx.Assignment (ConstrainedShape m) argTypes)
+argConstraints = lens _argConstraints (\s v -> s {_argConstraints = v})
+
+globalConstraints :: Simple Lens (Constraints m globalTypes) (Map L.Symbol (Some (ConstrainedShape m)))
+globalConstraints = lens _globalConstraints (\s v -> s {_globalConstraints = v})
+
+relationalConstraints :: Simple Lens (Constraints m argTypes) [RelationalConstraint m argTypes]
+relationalConstraints = lens _relationalConstraints (\s v -> s {_relationalConstraints = v})
+
+-- | The minimal set of constraints on this 'Ctx.Assignment' of types. For each
+-- input type, the generated 'ConstrainedShape' has the minimal memory footprint
+-- (all pointers are not assumed to point to allocated memory) and an empty list
+-- of 'Constraint'.
+emptyArgConstraints ::
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  Ctx.Assignment (ConstrainedShape m) argTypes
+emptyArgConstraints argTypes =
+  fmapFC
+    ( \argType ->
+        ConstrainedShape
+          (fmapFC (\_ -> Compose []) $ Shape.minimal argType)
+    )
+    argTypes
+
+-- | The minimal set of constraints on a program state: no constraints
+-- whatsoever on globals, and only the minimal constraints (see
+-- 'emptyArgConstraints') on the function arguments.
+emptyConstraints ::
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  Constraints m argTypes
+emptyConstraints argTypes =
+  Constraints
+    { _argConstraints = emptyArgConstraints argTypes,
+      _globalConstraints = Map.empty,
+      _relationalConstraints = []
+    }
+
+_oneArgumentConstraint ::
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  Ctx.Index argTypes inTy ->
+  ConstrainedShape m inTy ->
+  Constraints m argTypes
+_oneArgumentConstraint argTypes idx shape =
+  emptyConstraints argTypes & argConstraints . ixF' idx .~ shape
+
+ppConstraints :: Constraints m argTypes -> Doc Void
+ppConstraints (Constraints args globCs relCs) =
+  PP.vsep
+    ( catMaybes
+        [ if Ctx.sizeInt (Ctx.size args) == 0
+            then Nothing
+            else
+              Just $
+                nestSep
+                  ( PP.pretty "Arguments:" :
+                    toListFC
+                      (Shape.ppShape ppConstraints' . getConstrainedShape)
+                      args
+                  ),
+          if Map.size globCs == 0
+            then Nothing
+            else
+              Just $
+                nestSep
+                  [ PP.pretty "Globals: TODO"
+                  ],
+          if null relCs
+            then Nothing
+            else
+              Just $
+                nestSep
+                  [ PP.pretty "Relational constraints: TODO"
```

Comment:
Seems like this is probably an important TODO...

Reply:
The tool doesn't currently generate any relational constraints, so I don't feel like it's super important to tackle right now. I could add a comment to that effect?

Reply:
yeah, definitely worth pointing out, in that case

Reply:
Comments added!